src/passes/MergeFunctions.cpp

```diff
@@ -0,0 +1,591 @@
+/*
+
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Merge similar functions that only differs constant values (like immediate
+// operand of const and call insts) by parameterization.
+// Performing this pass at post-link time can merge more functions across
+// objects. Inspired by Swift compiler's optimization which is derived from
+// LLVM's one:
+// https://github.com/apple/swift/blob/main/lib/LLVMPasses/LLVMMergeFunctions.cpp
+// https://github.com/llvm/llvm-project/blob/main/llvm/docs/MergeFunctions.rst
+//
+// e.g.
+//
+//  Before:
+//    (func $big-const-42 (result i32)
+//      [[many instr 1]]
+//      (i32.const 44)
+//      [[many instr 2]]
+//    )
+//    (func $big-const-43 (result i32)
+//      [[many instr 1]]
+//      (i32.const 45)
+//      [[many instr 2]]
+//    )
+//  After:
+//    (func $byn$mgfn-shared$big-const-42 (result i32)
+//      [[many instr 1]]
+//      (local.get $0)
+//      [[many instr 2]]
+//    )
+//    (func $big-const-42 (result i32)
+//      (call $byn$mgfn-shared$big-const-42
+//       (i32.const 42)
+//      )
+//    )
+//    (func $big-const-43 (result i32)
+//      (call $byn$mgfn-shared$big-const-42
+//       (i32.const 43)
+//      )
+//    )
+
+#include "ir/hashed.h"
+#include "ir/manipulation.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/utils.h"
+#include "opt-utils.h"
+#include "pass.h"
+#include "support/hash.h"
+#include "support/utilities.h"
+#include "wasm.h"
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <ostream>
+#include <variant>
+#include <vector>
+
+namespace wasm {
+
+struct ConstValueDiff : public std::monostate {};
+struct ConstCalleeDiff : public std::monostate {};
+
+using ConstDiff =
+  std::variant<ConstValueDiff, Literals, ConstCalleeDiff, std::vector<Name>>;
```

Comment:
It's a little hard to read this code as it is not obvious to me what `ConstDiff` is meant to be. A comment explaining it would avoid needing to infer that from the usage below. Same for the monostates above.